### PR TITLE
First implementation of actual GCI logic.

### PR DESCRIPTION
**Pull request by danielfireman**
Still missing some bits and testing. Will be adding them on next pull requests. Relates to #6

---
This change is [Reviewable](https://reviewable.io/reviews/danielfireman/gci/9)

### CODE REVIEW

**Comment by emersoncantalice** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9#-:-KbHetc9MoirztehA3Cn:bnfp4nl)*

**Comment by emersoncantalice** [comment]
:lgtm:

---

Review status: 0 of 3 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9#-:-KbHexpRiML6a5F0i-Go:bnfp4nl)*

File: python/gci-flask/example/server.py [hunk 1]
```diff
@@ -10,12 +10,12 @@
 
 # Relative import for GCI.
 # TODO(danielfireman): Improve this importing hack. Most likely the best option is to upload to PIP.
-sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "../gci.py")))
+sys.path.append(os.path.abspath(os.path.join(os.getcwd(), "../flask_gci.py")))
```
Comment (barroca) [inline python/gci-flask/example/server.py]: I see the TODO, but it is very important to have this before using it in production. :)

File: python/gci-flask/example/server.py [hunk 2]
```diff
@@ -10,12 +10,12 @@
...
 import flask_gci
 
 # Adding GCI interceptor to the app.
-app.before_request(flask_gci.before)
-app.after_request(flask_gci.after)
+app.before_request(flask_gci.before_request)
+app.teardown_request(flask_gci.after_response)
```
Comment (barroca) [inline python/gci-flask/example/server.py]: shouldn't the function name be after_requrest?

File: python/gci/gci.py [hunk 3]
```diff
@@ -1,23 +1,75 @@
...
+        # GCI is only going to update virtual memory consumption first time it is called
+        # after GC collection.
+        if not self._virt:
+            self._virt = virt
+
+        # TODO(danielfireman): Use sampling window
+        # TODO(danielfireman): Disable GCI and enable automatic GC if there is a memory leak.
+        # The rationale here is to be no worse than automatic GC in extreme cases (like
+        # coding bugs)
+        if rss / self._virt > _SHEDDING_THRESHOLD:
```
Comment (barroca) [inline python/gci/gci.py]: if for some reason _virt becomes 0 it will raise error.

File: python/gci/gci.py [hunk 4]
```diff
@@ -1,23 +1,75 @@
...
-    def after(self, shed_response):
-        # TODO(danielfireman): Implement this method
-        pass
+    def _get_meminfo(self, filename=None):
+        # For Linux we can use information from the proc filesystem. We use
+        # '/proc/statm' as it easier to parse than '/proc/status' file.
+        filename = filename or _STATM_FILENAME
+        with open(filename, 'r') as fp:
+            statm = fp.read().split()
+            return float(statm[0]) * _PAGE_SIZE, float(statm[1]) * _PAGE_SIZE
```
Comment (barroca) [inline python/gci/gci.py]: don't we need to close the fp?

**Comment by danielfireman** [comment]
Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions.

---

*[python/gci/gci.py, line 54 at r1](https://reviewable.io:443/reviews/danielfireman/gci/9#-KbH04zLV7vnveHjqN4K-r1-54:-KbKjXmOt_NMEhCc_bMK:b-896fix) ([raw file](https://github.com/danielfireman/gci/blob/bbc2d72d48ed07cb3ae79d1a2abaa9efbfc66fe4/python/gci/gci.py#L54)):*
<details><summary><i>Previously, barroca (Leonardo da Mata) wrote…</i></summary><blockquote>

if for some reason _virt becomes 0 it will raise error.
</blockquote></details>

Done.

---

*[python/gci/gci.py, line 75 at r1](https://reviewable.io:443/reviews/danielfireman/gci/9#-KbH04zLV7vnveHjqN4K-r1-75:-KbKn_JwLE3gK4Q1pbqd:bw1fnbg) ([raw file](https://github.com/danielfireman/gci/blob/bbc2d72d48ed07cb3ae79d1a2abaa9efbfc66fe4/python/gci/gci.py#L75)):*
<details><summary><i>Previously, barroca (Leonardo da Mata) wrote…</i></summary><blockquote>

don't we need to close the fp?
</blockquote></details>

The with statement takes of this:

http://effbot.org/zone/python-with-statement.htm

---

*[python/gci-flask/example/server.py, line 18 at r1](https://reviewable.io:443/reviews/danielfireman/gci/9#-KbH04zLV7vnveHjqN4I-r1-18:-KbKnrZ7d8B1-PsqXwnA:b81s6e2) ([raw file](https://github.com/danielfireman/gci/blob/bbc2d72d48ed07cb3ae79d1a2abaa9efbfc66fe4/python/gci-flask/example/server.py#L18)):*
<details><summary><i>Previously, barroca (Leonardo da Mata) wrote…</i></summary><blockquote>

shouldn't the function name be after_requrest?
</blockquote></details>

Not really. The reationale behind this option is that flask's after_request allow for handler to change the response, so it had not sent the response yet. Running GC at this time would lead to an increase in response time (which is something we don't want).

So, after some research I found the  after_teardown method, which should be executed after the response has been sent (as per http://flask.pocoo.org/docs/0.12/reqcontext/#teardown-callbacks).

As stated in the module comment, the working logic of GCI in Cpython + Linux could be very simple (changing to be like this):

1. Pre-request: if it is the right time, check memory usage and if it is time to do GC return the estimated unavailability duration. This is only request that needs to be shed and it is only needed because we need to signal to LBs or any other clients how long the service will be unavailable doing GC.
2. Execute or shed the request
3. Do GC
4. Enjoy


:dancer: 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9)*

**Comment by barroca** [comment]
:lgtm:

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9#-:-KbKvx10ru4n8Gj5MWPi:b4xj4y9)*

**Comment by danielfireman** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/gci-flask/example/server.py, line 13 at r1](https://reviewable.io:443/reviews/danielfireman/gci/9#-KbH04zLV7vnveHjqN4I-r1-13:-KbLf33gW4KixFD6UyTT:b-896fix) ([raw file](https://github.com/danielfireman/gci/blob/bbc2d72d48ed07cb3ae79d1a2abaa9efbfc66fe4/python/gci-flask/example/server.py#L13)):*
<details><summary><i>Previously, barroca (Leonardo da Mata) wrote…</i></summary><blockquote>

I see the TODO, but it is very important to have this before using it in production. :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9)*

**Comment by danielfireman** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[python/gci-flask/example/server.py, line 13 at r1](https://reviewable.io:443/reviews/danielfireman/gci/9#-KbH04zLV7vnveHjqN4I-r1-13:-KbLfG-NloaW3inSFnZj:b-896fix) ([raw file](https://github.com/danielfireman/gci/blob/bbc2d72d48ed07cb3ae79d1a2abaa9efbfc66fe4/python/gci-flask/example/server.py#L13)):*
<details><summary><i>Previously, danielfireman (Daniel Fireman) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9)*

**Comment by danielfireman** [comment]
Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/danielfireman/gci/9)*

**Pull request merged**